test/unit/maybe_callback.test.js

```diff
@@ -0,0 +1,196 @@
+// @ts-nocheck
+'use strict';
+
+const sinon = require('sinon');
+const { expect } = require('chai');
+
+const mongodbDriver = require('mongodb');
+const mongodbLegacy = require('../..');
+const { MongoDBNamespace } = require('mongodb/lib/utils');
+const { asyncApi } = require('../tools/api');
+
+const classesToMethods = new Map(
+  asyncApi.map((api, _, array) => [
+    api.className,
+    new Set(array.filter(v => v.className === api.className))
+  ])
+);
+
+const iLoveJs = 'mongodb://iLoveJavascript';
+const client = new mongodbLegacy.MongoClient(iLoveJs);
+const db = new mongodbLegacy.Db(client, 'animals');
+const collection = new mongodbLegacy.Collection(db, 'pets');
+const namespace = MongoDBNamespace.fromString('animals.pets');
+
+const OVERRIDDEN_CLASSES_GETTER = new Map([
+  ['Admin', () => new mongodbLegacy.Admin(db)],
+  ['FindCursor', () => new mongodbLegacy.FindCursor(client, namespace)],
+  ['ListCollectionsCursor', () => new mongodbLegacy.ListCollectionsCursor(db, {})],
+  ['ListIndexesCursor', () => new mongodbLegacy.ListIndexesCursor(collection)],
+  ['AggregationCursor', () => new mongodbLegacy.AggregationCursor(client, namespace)],
+  ['ChangeStream', () => new mongodbLegacy.ChangeStream(client)],
+  ['GridFSBucket', () => new mongodbLegacy.GridFSBucket(db)],
+  ['Collection', () => new mongodbLegacy.Collection(db, 'pets')],
+  ['Db', () => new mongodbLegacy.Db(client, 'animals')],
+  ['MongoClient', () => new mongodbLegacy.MongoClient(iLoveJs)]
+]);
+
+function* generateTests() {
+  for (const [className, getInstance] of OVERRIDDEN_CLASSES_GETTER) {
+    /** @type {{ className: string; method: string; returnType: string; special?: boolean; }[]} */
+    let methods = classesToMethods.get(className);
+    if (methods == null) methods = [];
+    for (const { method, special, possibleCallbackPositions } of methods) {
+      if (special === true) continue;
+      const apiName = `${className}.${method}`;
+      const instance = getInstance();
+      yield {
+        className,
+        method,
+        instance,
+        apiName,
+        possibleCallbackPositions,
+        makeStub: superPromise => {
+          return sinon.stub(mongodbDriver[className].prototype, method).returns(superPromise);
+        }
+      };
+    }
+  }
+}
+
+describe('Maybe Callback', () => {
+  afterEach(() => {
+    sinon.restore();
+  });
+
+  for (const {
+    apiName,
+    instance,
+    method,
+    possibleCallbackPositions,
+    makeStub
+  } of generateTests()) {
+    const functionLength = instance[method].length;
+
+    context(`${apiName}()`, () => {
+      for (const callbackPosition of possibleCallbackPositions != null
+        ? possibleCallbackPositions
+        : [1, 2]) {
+        if (callbackPosition === 2) {
+          if (functionLength < 2) continue;
+          if (
+            apiName.includes('forEach') ||
```

Comment:
Why are we excluding `forEach` methods here?  It looks like those are `AbstractCursor.forEach`, which do take an optional callback.  I also don't see these explicitly tested elsewhere in this PR?

Reply:
We shouldn't be, fixed!